### PR TITLE
locators improved

### PR DESCRIPTION
**Pull request by StevenMaMei**

### CODE REVIEW

**Review by holgiosalos — COMMENTED** [review]
@StevenMaMei I have thought better, I think that I gave wrong advice, the time should not and cannot be the metric to see if our selectors are good or not, the metric should be the maintainability of the tests, the less flaky is the test the better, that means that a locator should be a combination between speed and isolation, so, we should take care of the efficiency but the most important thing is choosing a locator which its probability of change during the time is low.

This image is taken from our slides about the UI Testing:
image


I will finish your review tomorrow.

File: src/page/product-added-modal.page.ts [hunk 1]
```diff
@@ -5,9 +5,10 @@ export class ProductAddedPage {
   private timeout: number = 5000;
 
   constructor () {
-    this.addToCartBtn = $('#center_column a.button.ajax_add_to_cart_button.btn.btn-default');
+    // in this way we can select another product in the list by modifying the nth-child
+    // also, a.button.ajax_add_to_cart_button is enough
+    this.addToCartBtn = $('#center_column ul li:nth-child(1) a.button.ajax_add_to_cart_button');
```
Comment (holgiosalos) [inline src/page/product-added-modal.page.ts]: I just figure out, you confused product-added-modal with product-list page objects. This should have the proceedToCheckout

File: src/page/product-list.page.ts [hunk 2]
```diff
@@ -5,7 +5,9 @@ export class ProductListPage {
   private timeout: number = 5000;
 
   constructor () {
-    this.proceedCheckoutBtn = $('[style*="display: block;"] .button-container > a');
+    // this way we ensure that we are selecting the .button-container
+    // within the layer_cart (it is safer because it is an id)
+    this.proceedCheckoutBtn = $('#layer_cart .button-container > a');
```
Comment (holgiosalos) [inline src/page/product-list.page.ts]: and this, should have the addProduct

File: src/page/summary-step.page.ts [hunk 3]
```diff
@@ -4,7 +4,8 @@ export class SummaryStepPage {
   private proceedCheckoutBtn: ElementFinder;
 
   constructor () {
-    this.proceedCheckoutBtn = $('.cart_navigation span');
+    // it is more meaningful if we use also an 'a'
+    this.proceedCheckoutBtn = $('.cart_navigation a span');
```
Comment (holgiosalos) [inline src/page/summary-step.page.ts]: I think that `.cart_navigation span` is the only combination in that page that points to the proceed to checkout button. Maybe, the a is redundant

File: src/page/sign-in-step.page.ts [hunk 4]
```diff
@@ -11,7 +11,8 @@ export class SignInPage {
   constructor (email: string, password: string) {
     this.emailBox = $('#email');
     this.passwordBox = $('#passwd');
-    this.submitBtn = $('#SubmitLogin > span');
+    // we can remove '> span'
+    this.submitBtn = $('#SubmitLogin');
```
Comment (holgiosalos) [inline src/page/sign-in-step.page.ts:15]: 👍

**Comment by StevenMaMei** [comment]
> @StevenMaMei I have thought better, I think that I gave wrong advice, the time should not and cannot be the metric to see if our selectors are good or not, the metric should be the maintainability of the tests, the less flaky is the test the better, that means that a locator should be a combination between speed and isolation, so, we should take care of the efficiency but the most important thing is choosing a locator which its probability of change during the time is low.
> 
> This image is taken from our slides about the UI Testing:
> image
> 
> I will finish your review tomorrow.

Oh quite interesting. In my tests I was searching for something similar. At first, I tried to make the selectors more generally, that is to say, I made them in a way that they have low probability of change during the time. Otherwise, If I couldn't because the elements didn't have unique elements like ids, I made them meaningful for better understanding.

File: src/page/product-list.page.ts [hunk 5]
```diff
@@ -1,17 +1,18 @@
...
 export class ProductListPage {
-  private proceedCheckoutBtn: ElementFinder;
+  private addToCartBtn: ElementFinder;
   private timeout: number = 5000;
 
   constructor () {
-    this.proceedCheckoutBtn = $('[style*="display: block;"] .button-container > a');
+    // in this way we can select another product in the list by modifying the nth-child
+    // also, a.button.ajax_add_to_cart_button is enough
+    this.addToCartBtn = $('#center_column ul li:nth-child(1) a.button.ajax_add_to_cart_button');
```
Comment (holgiosalos) [inline src/page/product-list.page.ts:10]: I agree with you `a.button.ajax_add_to_cart_button` is enough

File: src/page/product-added-modal.page.ts [hunk 6]
```diff
@@ -1,17 +1,19 @@
...
 export class ProductAddedPage {
-  private addToCartBtn: ElementFinder;
+  private proceedCheckoutBtn: ElementFinder;
   private timeout: number = 5000;
 
   constructor () {
-    this.addToCartBtn = $('#center_column a.button.ajax_add_to_cart_button.btn.btn-default');
+    // this way we ensure that we are selecting the .button-container
+    // within the layer_cart (it is safer because it is an id)
+    this.proceedCheckoutBtn = $('#layer_cart .button-container > a');
```
Comment (holgiosalos) [inline src/page/product-added-modal.page.ts]: 👍 `a[title="Proceed to checkout"]` also would work. But I'm ok with this

File: src/page/summary-step.page.ts [hunk 7]
```diff
@@ -4,6 +4,7 @@ export class SummaryStepPage {
   private proceedCheckoutBtn: ElementFinder;
 
   constructor () {
+    // it is more meaningful if we use also an 'a'
     this.proceedCheckoutBtn = $('.cart_navigation span');
```
Comment (holgiosalos) [inline src/page/summary-step.page.ts]: I'm ok with leaving this without change, however, something more specific would be `.cart_navigation > a[title="Proceed to checkout"]`

File: src/page/address-step.page.ts [hunk 8]
```diff
@@ -4,7 +4,8 @@ export class AddresStepPage {
   private proceedToCheckoutBtn: ElementFinder;
 
   constructor () {
-    this.proceedToCheckoutBtn = $('#center_column > form > p > button > span');
+    // it could be clearer that we are searching the submit button
+    this.proceedToCheckoutBtn = $('#center_column form button');
```
Comment (holgiosalos) [inline src/page/address-step.page.ts]: I think that there is a better locator. Try to find one using a tag attribute.

File: src/page/shipping-step.page.ts [hunk 9]
```diff
@@ -6,7 +6,8 @@ export class ShippingStepPage {
 
   constructor () {
     this.agreeTermsCheckBox = $('#cgv');
-    this.proceedCheckoutBtn = $('#form > p > button > span');
+    // it could be clearer that we are searching the submit button
+    this.proceedCheckoutBtn = $('#form button');
```
Comment (holgiosalos) [inline src/page/shipping-step.page.ts]: This is acceptable. However, it doesn't change too much if we compare it with the original. You can use a better one

File: src/page/payment-step.page.ts [hunk 10]
```diff
@@ -4,7 +4,10 @@ export class PaymentStepPage {
   private payByBankBtn: ElementFinder;
 
   constructor () {
-    this.payByBankBtn = $('#HOOK_PAYMENT > div:nth-child(1) > div > p > a');
+    // can be shortened. In this way it is clearer that we
+    // are choosing the bankwire option, obviously it is
+    // a class, but it is better than the original selector
+    this.payByBankBtn = $('#HOOK_PAYMENT .bankwire');
```
Comment (holgiosalos) [inline src/page/payment-step.page.ts]: 👍

File: src/page/bank-payment.page.ts [hunk 11]
```diff
@@ -4,7 +4,8 @@ export class BankPaymentPage {
   private confirmOrderBtn: ElementFinder;
 
   constructor () {
-    this.confirmOrderBtn = $('#cart_navigation > button > span');
+    // we can remove span
+    this.confirmOrderBtn = $('#cart_navigation button');
```
Comment (holgiosalos) [inline src/page/bank-payment.page.ts]: I think that this doesn't make any difference with the original one, you can play a little bit more, try to use a different locator, for example, the text related with the button seems to be unique "I confirm my order"". Find a locator in where you can use that text.

**Review by holgiosalos — APPROVED** [review]
I think that you can find a better locator for the MenuContentPage and also for the OrderSummaryPage.

**Review by holgiosalos — CHANGES_REQUESTED** [review]
Sorry, I did approve but was a mistake

File: src/page/address-step.page.ts [hunk 12]
```diff
@@ -5,7 +5,7 @@ export class AddresStepPage {
 
   constructor () {
     // it could be clearer that we are searching the submit button
-    this.proceedToCheckoutBtn = $('#center_column form button');
+    this.proceedToCheckoutBtn = $('[name="processAddress"]');
```
Comment (holgiosalos) [inline src/page/address-step.page.ts:8]: 👍

File: src/page/bank-payment.page.ts [hunk 13]
```diff
@@ -1,11 +1,11 @@
-import { $, ElementFinder } from 'protractor';
+import { ElementFinder, element, by } from 'protractor';
 
 export class BankPaymentPage {
   private confirmOrderBtn: ElementFinder;
 
   constructor () {
     // we can remove span
-    this.confirmOrderBtn = $('#cart_navigation button');
+    this.confirmOrderBtn = element(by.cssContainingText('span', 'I confirm my order'));
```
Comment (holgiosalos) [inline src/page/bank-payment.page.ts:8]: 👍

Comment (holgiosalos) [reply src/page/bank-payment.page.ts:8]: Another option was: by.partialButtonText('I confirm my order') but, it yours also is good

File: src/page/payment-step.page.ts [hunk 14]
```diff
@@ -7,7 +7,7 @@ export class PaymentStepPage {
     // can be shortened. In this way it is clearer that we
     // are choosing the bankwire option, obviously it is
     // a class, but it is better than the original selector
-    this.payByBankBtn = $('#HOOK_PAYMENT .bankwire');
+    this.payByBankBtn = $('a[title="Pay by bank wire"]');
```
Comment (holgiosalos) [inline src/page/payment-step.page.ts:10]: 👍

File: src/page/product-added-modal.page.ts [hunk 15]
```diff
@@ -7,7 +7,7 @@ export class ProductAddedPage {
   constructor () {
     // this way we ensure that we are selecting the .button-container
     // within the layer_cart (it is safer because it is an id)
-    this.proceedCheckoutBtn = $('#layer_cart .button-container > a');
+    this.proceedCheckoutBtn = $('a[title="Proceed to checkout"]');
```
Comment (holgiosalos) [inline src/page/product-added-modal.page.ts:10]: 👍

File: src/page/shipping-step.page.ts [hunk 16]
```diff
@@ -7,7 +7,7 @@ export class ShippingStepPage {
   constructor () {
     this.agreeTermsCheckBox = $('#cgv');
     // it could be clearer that we are searching the submit button
-    this.proceedCheckoutBtn = $('#form button');
+    this.proceedCheckoutBtn = $('[name="processCarrier"]');
```
Comment (holgiosalos) [inline src/page/shipping-step.page.ts:10]: 👍

File: src/page/summary-step.page.ts [hunk 17]
```diff
@@ -4,8 +4,8 @@ export class SummaryStepPage {
   private proceedCheckoutBtn: ElementFinder;
 
   constructor () {
-    // it is more meaningful if we use also an 'a'
-    this.proceedCheckoutBtn = $('.cart_navigation span');
+    // it is better to use attributes
+    this.proceedCheckoutBtn = $('.cart_navigation a[title="Proceed to checkout"]');
```
Comment (holgiosalos) [inline src/page/summary-step.page.ts:8]: 👍

**Review by holgiosalos — COMMENTED** [review]
@StevenMaMei don't forget improve locators for: `MenuContentPage` and `OrderSummaryPage`

**Comment by StevenMaMei** [comment]
> @StevenMaMei don't forget improve locators for: `MenuContentPage` and `OrderSummaryPage`

Holmes mira que en el de menu-content-page hay que utilizar  li:nth-child(3), porque el primer "li" también tiene un link con title T-shirts

File: src/page/order-summary.page.ts [hunk 18]
```diff
@@ -1,10 +1,10 @@
-import { $, ElementFinder } from 'protractor';
+import { by, element, ElementFinder } from 'protractor';
 
 export class OrderSummaryPage {
   private confirmationText: ElementFinder;
 
   constructor () {
-    this.confirmationText = $('#center_column > div > p > strong');
+    this.confirmationText = element(by.cssContainingText('strong', 'Your order on My Store is complete.'));
```
Comment (holgiosalos) [inline src/page/order-summary.page.ts]: este selector está ligado a que la orden siempre se complete o termine de manera exitosa, o que el título siempre sea el que especificas. Este locator debería funcionar bien: `$('.cheque-indent')`

File: src/page/menu-content.page.ts [hunk 19]
```diff
@@ -4,7 +4,7 @@ export class MenuContentPage {
   private tShirtMenu: ElementFinder;
 
   constructor () {
-    this.tShirtMenu = $('#block_top_menu > ul > li:nth-child(3) > a');
+    this.tShirtMenu = $('#block_top_menu li:nth-child(3) a[title="T-shirts"]');
```
Comment (holgiosalos) [inline src/page/menu-content.page.ts:7]: Una alternativa puede ser: `#block_top_menu > ul > li > a[title="T-shirts"]'`

Comment (StevenMaMei) [reply src/page/menu-content.page.ts:7]: No, ese no funciona, ya lo había probado
https://github.com/StevenMaMei/protractor-workshop-2020/pull/14#issuecomment-665818129

Comment (holgiosalos) [reply src/page/menu-content.page.ts:7]: Que raro, solo me aparece uno. Igual dejalo asi, tenemos un buen resultado final
image

**Review by holgiosalos — APPROVED** [review]

**Pull request merged**